compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -1924,3 +1924,9 @@ fn determine_place_ancestry_relation(
         PlaceAncestryRelation::Divergent
     }
 }
+
+/// Precise capture is enabled if the feature gate `capture_disjoint_fields` is enabled or if
+/// user is using Rust Edition 2021 or higher
+fn enable_precise_capture(tcx: TyCtxt<'_>) -> bool {
+    tcx.features().capture_disjoint_fields || tcx.sess.rust_2021()
```

Comment:
Actually we should be using [`span.rust_2021()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.rust_2021), so this function needs to take a `Span` argument. We really want, I think, the span of the "closure header" (the `|...|` part), but I think if we just pass in the `span` of the closure overall, that will do.

The reason for tying the edition to the span is because macros may inject code from other editions. We should probably write some tests for that, too.

Reply:
The span of `|..| body` is made with `lo.to(body.span)`, where `lo` is the span of the first token (`move` or `async` or `static` or `||` or `|`). The `Span::to` function in most cases carries the edition of the left hand side, but the code has multiple FIXMEs and the behaviour when macros and multiple contexts are involved is not very obvious at all:

https://github.com/rust-lang/rust/blob/75ed34223adb02c7bbfa5ae5e62c48952ff29914/compiler/rustc_span/src/lib.rs#L657-L671

Reply:
Good enough for now